debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.base;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.function.Supplier;
+
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.config.ConfigurationDefaults;
+import io.debezium.time.Temporals;
+import io.debezium.util.Clock;
+import io.debezium.util.LoggingContext;
+import io.debezium.util.LoggingContext.PreviousContext;
+import io.debezium.util.Metronome;
+import io.debezium.util.Threads;
+import io.debezium.util.Threads.Timer;
+
+/**
+ * A queue which serves as handover point between producer threads (e.g. MySQL's
+ * binlog reader thread) and the Kafka Connect polling loop.
+ * <p>
+ * The queue is configurable in different aspects, e.g. its maximum size and the
+ * time to sleep (block) between two subsequent poll calls. See the
+ * {@link Builder} for the different options. The queue applies back-pressure
+ * semantics, i.e. if it holds the maximum number of elements, subsequent calls
+ * to {@link #enqueue(Object)} will block until elements have been removed from
+ * the queue.
+ * <p>
+ * If an exception occurs on the producer side, the producer should make that
+ * exception known by calling {@link #producerFailure} before stopping its
+ * operation. Upon the next call to {@link #poll()}, that exception will be
+ * raised, causing Kafka Connect to stop the connector and mark it as
+ * {@code FAILED}.
+ *
+ * @author Gunnar Morling
+ *
+ * @param <T>
+ *            the type of events in this queue. Usually {@link SourceRecord} is
+ *            used, but in cases where additional metadata must be passed from
+ *            producers to the consumer, a custom type wrapping source records
+ *            may be used.
+ */
+public class ChangeEventQueue<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChangeEventQueue.class);
+
+    private final Duration pollInterval;
+    private final int maxBatchSize;
+    private final BlockingQueue<T> queue;
+    private final Metronome metronome;
+    private final Supplier<PreviousContext> loggingContextSupplier;
+
+    private volatile Throwable producerFailure;
+
+    private ChangeEventQueue(Duration pollInterval, int maxQueueSize, int maxBatchSize, Supplier<LoggingContext.PreviousContext> loggingContextSupplier) {
```

Comment:
Can we have lifecycle and events listeners available? Something like
* onQueueStarted
* onPollCompleted
* onFaillure
etc.?

Reply:
Yes, good idea. But I'd do it when the need arises. Let's not build things in advance without seeing clear requirements.